sycl/source/detail/program_impl.cpp

```diff
@@ -358,8 +358,16 @@ void program_impl::create_cl_program_with_source(const string_class &Source) {
   const char *Src = Source.c_str();
   size_t Size = Source.size();
   const detail::plugin &Plugin = getPlugin();
-  Plugin.call<PiApiKind::piclProgramCreateWithSource>(
-      MContext->getHandleRef(), 1, &Src, &Size, &MProgram);
+  RT::PiResult Err =
+      Plugin.call_nocheck<PiApiKind::piclProgramCreateWithSource>(
+          MContext->getHandleRef(), 1, &Src, &Size, &MProgram);
+
+  if (Plugin.getBackend() == cl::sycl::backend::level_zero &&
+      Err == PI_INVALID_OPERATION) {
+    throw feature_not_supported(
+        "piclProgramCreateWithSource is not supported in Level Zero",
```

Comment:
```suggestion
        "program::compile_with_source is not supported with Level Zero backend.",
```

We must not tell end users about PI interfaces. Also I think it would be good to redirect to the library that can be used to build OpenCL kernel for Level-Zero. @v-klochkov , which one is this?

Reply:
Shouldn't this error be emitted from Level Zero plug-in, rather than from DPC++ runtime?

Reply:
Plugins are C API. In order to throw a consistent error message this is better suited for SYCL RT, I think.

Reply:
> In order to throw a consistent error message this is better suited for SYCL RT, I think.

Right, but BE specific error detection logic must be located in corresponding plug-in. DPC++ runtime is designed to keep device specific logic in plug-ins and logic common for all back-ends in "runtime" library.
To follow this design we can separate error detection logic (device specific) from error reporting/conversion logic (common for all plug-ins).
E.g. plug-in can return an error code when "compile with source" feature is not supported (PI_INVALID_OPERATION? or some other error code) and DPC++ runtime can report this error via exception - "program::compile_with_source is not supported with" + Plugin.getBackend().to_string()" - this is a pseudo code to demonstrate the idea, not a real C++ program.

Basically code like this: 
```C++
if (Plugin.getBackend() == cl::sycl::backend::<some_backend>) {
...
}
```
is a red flag signaling that most likely it should be implemented in corresponding plug-in.

Reply:
That's right. We don't need to check the backend here in SYCL RT, every backend would return PI_INVALID_OPERATION of compilation is not supported, and the error message would be common for all such backends. @rbegam , please also change CUDA plugin to do the same, i.e. return an error: https://github.com/intel/llvm/blob/eab4791cf727ac889857d4723c5c9284489c9537/sycl/plugins/cuda/pi_cuda.cpp#L2653

Reply:
> Also I think it would be good to redirect to the library that can be used to build OpenCL kernel for Level-Zero. @v-klochkov , which one is this?

The class sycl::INTEL::online_compiler can generate IL/SPIR-V for a source program.